src/main.rs

```diff
@@ -71,20 +71,26 @@ fn main() {
     } else {
         (save_c, save_ir) = (false, false);
     }
-    let src_file = fs::read_to_string(args.infile).expect("ERROR: couldn't find source file");
-    let root = rascal_grammar::RootParser::new().parse(&src_file).unwrap();
-    // Perform semantic checks and type checking
-    let mut state = semantic::new_state(root);
-    state.build().unwrap();
-    if save_ir {
-        let serialized_ir = serde_json::to_string(&state.build_stack).expect("Serialization error");
-        let mut file =
-            File::create(ProgramState::IR_OUTPUT_FILENAME).expect("Cannot create IR file");
-        write!(&mut file, "{serialized_ir}").expect("Cannot write to IR file");
-    }
-
+    let src_file = fs::read_to_string(&args.infile).expect("ERROR: couldn't find source file");
+    let build_stack = if Path::new(&args.infile).extension().expect("Problem with filename") == "ir" {
+        // maybe we should make input error an error enum type
```

Comment:
Do you think we should have an error type for input file errors

Reply:
That's a great idea! Maybe we can create an enum representing a general class of error relating to the CLI commands, opening source files, etc.

I have a couple of these implemented for other modules (far from complete though). I'd take a look at [the error enum for the codegen](https://github.com/drewrip/rascal/blob/a64d8b2896c139b1299a74458484712ea34aa687/src/codegen.rs#L10) module as a reference. And some [examples](https://github.com/drewrip/rascal/blob/a64d8b2896c139b1299a74458484712ea34aa687/src/backends/c.rs#L100) of how they get used.